bench/workgen/workgen.cpp

```diff
@@ -353,42 +359,91 @@ gen_random_table_name(char *name, workgen_random_state volatile *rand_state)
  * fails.
  */
 int
-WorkloadRunner::create_table(WT_SESSION *session, const std::string &config, const std::string &uri,
-  const std::string &mirror_uri, const bool is_base)
+WorkloadRunner::create_table(
+  WT_SESSION *session, const std::string &config, const std::string &uri, bool mirror_enabled)
 {
-    // Check if a table with this name already exists. Return if it does. Use a shared lock to
-    // read the dynamic table structure.
+    const std::string mirror_uri(mirror_enabled ? uri + _workload->options.mirror_suffix : "");
+
+    // Return if a table with the same name already exists. If mirror is enabled, make sure it does
+    // not exist either.
     ContextInternal *icontext = _workload->_context->_internal;
     {
+        // Use a shared lock to read the dynamic table structure.
         const std::shared_lock lock(*icontext->_dyn_mutex);
         if (icontext->_tint.count(uri) > 0 || icontext->_dyn_tint.count(uri) > 0)
             return EEXIST;
+        if (mirror_enabled &&
+          (icontext->_tint.count(mirror_uri) > 0 || icontext->_dyn_tint.count(mirror_uri) > 0))
+            return EEXIST;
     }
 
-    // Create the table.
-    WT_DECL_RET;
-    if ((ret = session->create(session, uri.c_str(), config.c_str())) != 0) {
-        if (ret != EBUSY)
-            THROW("Failed to create table '" << uri << "'.");
+    /*
+     * The config has to be updated with the following information:
+     *  - The name of the table's mirror if mirroring is enabled.
+     *  - If this table is a base table or a mirror.
+     */
+    std::string base_config(config);
+    if (mirror_enabled)
+        base_config += "," + MIRROR_TABLE_APP_METADATA;
+    else
+        base_config += "," + BASE_TABLE_APP_METADATA + "true\"";
+
+    // When mirror is enabled, create the mirror first and the base. If we create the base first,
+    // threads may start working on the base while the mirror is not fully created.
+    if (mirror_enabled) {
+        const std::string config_tmp(base_config + uri + "," + BASE_TABLE_APP_METADATA + "false\"");
+        int ret = session->create(session, mirror_uri.c_str(), config_tmp.c_str());
+        if (ret != 0) {
+            VERBOSE(*_workload, "Failed to create mirror table '" << mirror_uri << "'");
+            return ret;
+        }
+    }
+
+    // Create the base table.
+    std::string config_tmp(base_config);
+    if (mirror_enabled)
+        config_tmp += mirror_uri + "," + BASE_TABLE_APP_METADATA + "true\"";
```

Comment:
We set ```BASE_TABLE_APP_METADATA + "true\"``` earlier on line 389 when mirror is not enabled. We set here when mirror is enabled. Unless there is need to create the mirror successfully before creating the base config string we can merge the blocks for config string generation together.

There are 3 sections of if - else around mirror enabled, it is a bit difficult to understand the configuration generation. I think we can simplify these even if we have to duplicate the string generation code a bit. If you need brainstorming around this, we can talk specifics offline.

Reply:
I agree, it was confusing.
I tried to make it better in fd735db4ffbe942b8ef28adb390987fe497a054c, let me know what you think.